scripts/swap-beacon/src/index.ts

```diff
@@ -0,0 +1,308 @@
+import { multicallAbi } from "./abi";
+import yargs from "yargs";
+import { hideBin } from "yargs/helpers";
+
+export class SwapBeaconError extends Error {}
+
+export function getSwapAdapterConfig(chainId: string) {
+  const swapAdapterConfig = SWAP_ADAPTER_CONFIGS[chainId];
+  if (!swapAdapterConfig) {
+    throw new SwapBeaconError(
+      `Opportunity adapter config not found for chain id: ${chainId}`
+    );
+  }
+  return swapAdapterConfig;
+}
+
+export class SwapBeacon {
+  private client: Client;
+  private adapters: Adapter[];
+  private chainIds: string[];
+
+  constructor(endpoint: string, _chainIds: string[]) {
+    this.client = new Client(
+      {
+        baseUrl: endpoint,
+      },
+      undefined,
+      this.opportunityHandler.bind(this)
+    );
+    this.chainIds = _chainIds;
+    this.adapters = [new OdosAdapter()];
+  }
+
+  private getOptimalAdapter(
+    tokenIn: Address,
+    tokenOut: Address
+  ): Adapter | undefined {
+    if (tokenIn.toLowerCase() === tokenOut.toLowerCase()) {
+      return undefined;
+    }
+    return this.adapters[0];
+  }
+
+  private makeMulticallCalldata(
+    opportunity: Opportunity,
+    swapsSell: ExtendedTargetCall[],
+    swapsBuy: ExtendedTargetCall[],
+    sellTokens: TokenAmount[],
+    buyTokens: TokenAmount[]
+  ): Hex {
+    const originalTargetCall = {
+      targetContract: opportunity.targetContract,
+      targetCalldata: opportunity.targetCalldata,
+      targetCallValue: opportunity.targetCallValue,
+      tokensToSend: opportunity.sellTokens.map((tokenAmount) => ({
+        tokenAmount: tokenAmount,
+        destination: opportunity.targetContract,
+      })),
+    };
+    const swapsSellTargetCalls = swapsSell.map((swap) => ({
+      targetContract: swap.targetContract,
+      targetCalldata: swap.targetCalldata,
+      targetCallValue: swap.targetCallValue,
+      tokensToSend: swap.tokensToSend,
+    }));
+    const swapsBuyTargetCalls = swapsBuy.map((swap) => ({
+      targetContract: swap.targetContract,
+      targetCalldata: swap.targetCalldata,
+      targetCallValue: swap.targetCallValue,
+      tokensToSend: swap.tokensToSend,
+    }));
+    const multicallTargetCalls = [
+      ...swapsSellTargetCalls,
+      originalTargetCall,
+      ...swapsBuyTargetCalls,
+    ];
+
+    return encodeFunctionData({
+      abi: [multicallAbi],
+      args: [[sellTokens, buyTokens, multicallTargetCalls]],
+    });
+  }
+
+  private createSwapOpportunity(
+    opportunity: Opportunity,
+    base: Address,
+    swapsSell: ExtendedTargetCall[],
+    swapsBuy: ExtendedTargetCall[]
+  ): Opportunity | undefined {
+    const targetContract =
+      SWAP_ADAPTER_CONFIGS[opportunity.chainId].multicallAdapter;
+    const targetCallValue = BigInt(
+      BigInt(
+        swapsSell.reduce(
+          (prev, curr) => BigInt(prev + curr.targetCallValue),
+          0n
+        ) +
+          swapsBuy.reduce(
+            (prev, curr) => BigInt(prev + curr.targetCallValue),
```

Comment:
no need to wrap around BigInt

Reply:
On my local it won't work without wrapping.
I'll use docker and test if I can get rid of these bigint wrapping

Reply:
It was a bug